Proposing `package:typed_sql` (#272)
google/dart-neats: typed_sql/doc/08_transactions.md

> This document shows how to do transactions with `package:typed_sql`.
> Examples throughout this document assume a database schema defined as
> follows:

Reviewer: ```suggestion
# Transactions

Examples assume a database schema defined as
follows:
```

failed because the account wasn't found or because the balance was insufficient.

```dart bank_test.dart#transfer-100-with-conditional-update
await db.transact(() async {
  // Withdraw 100 from account 0000-01
  final newBalance = await db.accounts
      .byAccountNumber('0001')
      .where((a) => a.balance >= literal(100))
      .update((a, set) => set(balance: a.balance - literal(100)))
      .returning((a) => (a.balance,))
      .executeAndFetch();
  if (newBalance == null) {
    print('account not found or insufficient balance');
    return;
  }

  // Deposit 100 to account 0000-02
  await db.accounts
      .byAccountNumber('0002')
      .update((a, set) => set(balance: a.balance + literal(100)))
      .execute();
});
```

In practice, it's probably more natural to lookup the balance prior to making
the update. This incurs additional database round-trips, and using a database
configured without a strict [transaction isolation][1] this could lead to
inconsistencies.

## Nested transactions
While SQL doesn't have nested transactions, SQL does have `SAVEPOINT`s which can
be used to facilitate nested transactions. Using `package:typed_sql` you can do

> nested invocations of `db.transact()`. The outer most invocation will create a

Reviewer: ```suggestion
nested invocations of `db.transact()`. The outermost invocation will create a
```

> and transaction, and the nested invocations will create a unique `SAVEPOINT`

Reviewer: ```suggestion
transaction, and the nested invocations will create a unique `SAVEPOINT`
```

      await db.accounts
          .insert(
            accountNumber: literal('0002'),
            balance: literal(100),
          )
          .execute();
    });
  } on TransactionAbortedException {
    await db.accounts
        .byAccountNumber('0002')
        .update((a, set) => set(balance: a.balance + literal(100)))
        .execute();
  }
});
```

> [!WARNING]
> Nested transaction can be hard to use correctly, often you're probably better
> off trying to avoid nested transaction. In the example above, it would
> probably better to use two separate transactions, first attempting to update
> and if that fails (or there is nothing to update), then trying to insert a
> new row in the accounts table.

## Pitfalls with transactions
There are a few limitations and pitfalls when using transactions with
`package:typed_sql`. The most important thing to remember to always `await`
operations and transactions, in particularly operations inside transactions.

The `db.transact(cb)` method will start a transaction and execute the callback
`cb` inside a [Zone], such that all operations using `db` happens inside the
transaction. When the callback `cb` completes (or rather the future `cb`

> returns is completes), the transaction is committed (or rolledback if `cb`

Reviewer: ```suggestion
returns completes), the transaction is committed (or rolled back if `cb`
```

> threw an `Exception`). So if we were to additional calls on `db` inside the

Reviewer: ```suggestion
threw an `Exception`). So if we were to make additional calls on `db` inside the
```

transaction zone, they would throw errors at runtime, example:

```dart
await db.transact(() async {
  await db.byAccountNumber('0001').delete().execute();
  //▲
  //└──────── Always add 'await' as illustrated here.

  /***/ db.byAccountNumber('0002').delete().execute();
  //▲

>   //└─────── Missing await here is VERY BAD, may cause a runtime error!

Author: Done

});
```

Even if you don't care about whether or not the operation is successful,
you must **always `await` operations in transactions**.
Similarly, you must also `await` nested transactions.

Furthermore, to prevent deadlocks in transaction callbacks `Query<T>.stream()`
will not pause on back-pressure inside transactions. This means that
_inside a transaction_ it doesn't matter if you use `Query<T>.stream()` or
`Query<T>.fetch()`. In both cases, all the result-rows may be buffered in

> memory.

Author: then it'd have to be runtime error, that sucks, buffering is the alternative, which should be fine.

It's not liek streaming is particularly attractive, when you have `.fetch`.
